server/src/main/java/org/elasticsearch/action/admin/cluster/configuration/AddVotingConfigExclusionsRequest.java

```diff
@@ -56,32 +62,72 @@ public AddVotingConfigExclusionsRequest(String[] nodeDescriptions) {
+        // TODO should this be removed in the latest version where nodeIds and nodeNames are used?
         nodeDescriptions = in.readStringArray();
+        // TODO which version to use here?
+        if (in.getVersion() == Version.V_EMPTY) {
+            nodeIds = in.readStringArray();
+            nodeNames = in.readStringArray();
+        }
         timeout = in.readTimeValue();
     }
 
     Set<VotingConfigExclusion> resolveVotingConfigExclusions(ClusterState currentState) {
         final DiscoveryNodes allNodes = currentState.nodes();
-        final Set<VotingConfigExclusion> resolvedNodes = Arrays.stream(allNodes.resolveNodes(nodeDescriptions))
+        Set<VotingConfigExclusion> allProcessedNodes = null;
+
+        if (nodeDescriptions.length >= 1) {
+            allProcessedNodes = Arrays.stream(allNodes.resolveNodes(nodeDescriptions)).map(allNodes::get)
+                                    .filter(DiscoveryNode::isMasterNode).map(VotingConfigExclusion::new).collect(Collectors.toSet());
+
+            if (allProcessedNodes.isEmpty()) {
+                throw new IllegalArgumentException("add voting config exclusions request for " + Arrays.asList(nodeDescriptions)
+                    + " matched no master-eligible nodes");
+            }
+        }
+        else {
+            DiscoveryNodes.NodeResolutionResults nodeResolutionResults;
+            Set<VotingConfigExclusion> resolvedNodes;
+            Set<VotingConfigExclusion> unresolvedNodes;
+
+            if (nodeIds.length >= 1) {
+                nodeResolutionResults = allNodes.resolveNodesExact(true, nodeIds);
+
+                unresolvedNodes = Arrays.stream(nodeResolutionResults.getUnresolvedNodes())
+                    .map(nodeId -> new VotingConfigExclusion(nodeId, "")).collect(Collectors.toSet());
+            }
+            else {
+                nodeResolutionResults = allNodes.resolveNodesExact(false, nodeNames);
+
+                unresolvedNodes = Arrays.stream(nodeResolutionResults.getUnresolvedNodes())
+                    .map(nodeName -> new VotingConfigExclusion("", nodeName)).collect(Collectors.toSet());
+            }
+
+            resolvedNodes = Arrays.stream(nodeResolutionResults.getResolvedNodes())
                 .map(allNodes::get).filter(DiscoveryNode::isMasterNode).map(VotingConfigExclusion::new).collect(Collectors.toSet());
 
-        if (resolvedNodes.isEmpty()) {
-            throw new IllegalArgumentException("add voting config exclusions request for " + Arrays.asList(nodeDescriptions)
-                + " matched no master-eligible nodes");
+            allProcessedNodes = Sets.newHashSet(Iterables.concat(resolvedNodes, unresolvedNodes));
+
+            if (allProcessedNodes.isEmpty()) {
+                throw new IllegalArgumentException("add voting config exclusions request for nodeIds " + Arrays.asList(nodeIds) +
+                    " or nodeNames " + Arrays.asList(nodeNames) + " matched no master-eligible nodes or absent nodes");
```

Comment:
Can this happen? I think only if both args are empty? In which case I think we can detect it sooner and throw a more helpful message.

Reply:
You are right, in the new approach this shouldn't happen anymore. This is more of a left-over from previous implementation. I've removed it.